http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/NonLeoMessageSubscriber.scala

```diff
@@ -0,0 +1,144 @@
+
+  private[monitor] def messageHandler(event: Event[NonLeoMessage]): F[Unit] = {
+    val traceId = event.traceId.getOrElse(TraceId("None"))
+    for {
+      now <- nowInstant[F]
+      implicit0(ev: Ask[F, AppContext]) <- F.pure(Ask.const[F, AppContext](AppContext(traceId, now, None)))
+      _ <- metrics.incrementCounter(event.msg.messageType)
+      res <- messageResponder(event.msg).attempt
+      _ <- res match {
+        case Left(e)  => logger.error(e)("Fail to process pubsub message") >> F.delay(event.consumer.nack())
+        case Right(_) => F.delay(event.consumer.ack())
+      }
+    } yield ()
+  }
+
+  private[monitor] def messageResponder(
+    message: NonLeoMessage
+  )(implicit traceId: Ask[F, AppContext]): F[Unit] = message match {
+    case m: DeleteKubernetesClusterMessage => handleDeleteKubernetesClusterMessage(m)
+    case m: NonLeoMessage.CryptoMining     => handleCryptoMiningMessage(m)
+  }
+
+  private[monitor] def handleDeleteKubernetesClusterMessage(msg: DeleteKubernetesClusterMessage)(
+    implicit ev: Ask[F, AppContext]
+  ): F[Unit] =
+    for {
+      ctx <- ev.ask
+      clusterId = msg.clusterId
+      _ <- kubernetesClusterQuery.markPendingDeletion(clusterId).transaction
+      _ <- gkeAlg
+      // TODO: Should we retry failures and with what RetryConfig? If all retries fail, send an alert?
+        .deleteAndPollCluster(DeleteClusterParams(msg.clusterId, msg.project))
+        .onError {
+          case _ =>
+            for {
+              _ <- logger.error(
+                s"An error occurred during clean-up of cluster ${clusterId} in project ${msg.project}. | trace id: ${ctx.traceId}"
+              )
+              _ <- kubernetesClusterQuery.updateStatus(clusterId, KubernetesClusterStatus.Error).transaction
+              // TODO: Create a KUBERNETES_CLUSTER_ERROR table to log the error message?
+              // TODO: Need mark the nodepool(s) as Error'ed too?
+            } yield ()
+        }
+    } yield ()
+
+  private[monitor] def handleCryptoMiningMessage(
+    msg: NonLeoMessage.CryptoMining
+  )(implicit ev: Ask[F, AppContext]): F[Unit] =
+    for {
+      ctx <- ev.ask
+      _ <- if (!msg.textPayload.contains(
+                 "CRYPTOMINING_DETECTED"
+               )) // needs to match https://github.com/broadinstitute/terra-cryptomining-security-alerts/blob/master/v2/main.go#L24
+        F.unit
+      else
+        for {
+          runtimeFromGoogle <- computeService.getInstance(msg.resource.labels.projectId,
+                                                          msg.resource.labels.zone,
+                                                          InstanceName(msg.resource.labels.instanceId.toString))
+          // We mark the runtime as Deleted, and delete the instance.
+          // If instance deletion fails for some reason, it will be cleaned up by resource-validator
+          _ <- runtimeFromGoogle.traverse { instance =>
+            for {
```

Comment:
Can we update a metric too? So we can make a stackdriver dashboard for this

Reply:
there's https://github.com/DataBiosphere/leonardo/pull/1711/files/ed6b2882d121c6b0f2f4f1f25764b2c5c3058eb8#diff-1ea743458d40ca65f6c9a7f7d9178ded5967d287a3113158f4e9f713b01d930bR43....are you thinking something more?

Reply:
Oh missed that.. that's probably fine then